README
keploy 0.7
===============

keploy - SSH Public Key Deplyment Utility
Copyright © 2007 Greg Swift
Written by: Greg Swift <greg.swift@nytefyre.net>
License: General Public License (GPL) v3

Description
===============

keploy is a utility to distribute and remove your public ssh key to a single
or large number of systems with the least amount of effort possible.

Usage
===============

* To deploy your default public key (~/.ssh/{id_rsa,id_dsa,identity}.pub) to all
  hosts in your known_hosts file:
     # keploy -k

* To deploy your default public key to host remote.host.com:
     # keploy remote.host.com

* To deploy your default public key to host remote.host.com and auto accept
  the ssh public key:
     # keploy -y remote.host.com

* To deploy your default public key to remote.host.com as user bob and
  enable Agent Forwarding:
     # keploy -A --user=bob remote.host.com

* To replace your old key (now ~/.ssh/id_rsa.old.pub) with your new default 
  public key onto all hosts in the file "~/myhosts":
     # keploy -c ~/.ssh/id_rsa.old.pub -f ~/myhosts

* Remove your public key from all boxes in your known_hosts file,
  and turn off Agent Forwarding:
     # keploy -r -k -A

How to report bugs
===============
Visit http://code.google.com/p/keploy and goto the 'Issues' tab


CHANGELOG
Release 0.7 - Nov 8th, 2012 - Greg Swift
- Added timeout to cli options
- Added autoaccept of unknown SSH public keys as cli option

Release 0.6 - Jun 22nd, 2010 - Greg Swift 
- Re-wrote the entire application to be module based
- Changed cli switch for old identity file from -c to -o
- Restructured project to utilize distutils
- Re-wrote layout of changelog file

Release 0.5 - Feb 4th, 2010 - Greg Swift
- Fix bug #1 - keploy properly error on inaccessible id file
- Added over_ride default value in getIdentity function
- Added CHANGELOG file

Release 0.4 - Mar 06, 2008 - Greg Swift
- First proper release of the code


keploy/__init__.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# keploy - SSH Public Key Deplyment Utility
# Copyright © 2007 Greg Swift gregswift@gmail.com
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os
import sys

class KeployError(Exception):
  def __init__(self, message, code=255):
    Exception.__init__(self, message, code)

"""
 Define Variables
"""
DEBUG = False
VERSION = "1.0"

PW_WARN = '\nNOTICE: You may be prompted for you password,\n'
PW_WARN += 'NOTICE: this is directly from the ssh client, not keploy\n'

SSH_BIN = os.popen('/usr/bin/which ssh 2> /dev/null').read().strip()
if SSH_BIN == '':
  raise KeployError, ('SSH binary not locatable by /usr/bin/which', 100)
elif not os.access(SSH_BIN, os.X_OK):
  raise KeployError, ('SSH Binary not executable by your user', 101)
SSH_DIR = '~/.ssh'
SSH_LOCAL_DIR = os.path.expanduser(SSH_DIR)
SSH_CONFIGS = ('/etc/ssh/ssh_config', os.path.join(SSH_DIR,'config'))
SSH_START_CALL = SSH_BIN+' -qq %s %s %s \''
SSH_OPTION_TIMEOUT = '-oConnectTimeout=%s'
SSH_OPTION_ACCEPTKEY = '-oStrictHostKeyChecking=no'
SSH_END_CALL = '\''
# Local Files
ID_FILES = (os.path.join(SSH_LOCAL_DIR, 'id_rsa.pub'),
    os.path.join(SSH_LOCAL_DIR, 'id_dsa.pub'),
    os.path.join(SSH_LOCAL_DIR, 'identity.pub'))
HOST_FILES = [os.path.join(SSH_LOCAL_DIR, 'known_hosts'),
    os.path.join(SSH_LOCAL_DIR, 'known_hosts2')]
# Files on remote host
TMP_FILE = os.path.join(SSH_DIR, 'keyploy.tmp')
AUTH_KEYS_FILE = os.path.join(SSH_DIR, 'authorized_keys')

"""
 Define Output Functions and cleanUp Function
"""
def standardOut(msg, on=True):
  """
  When quiet is not enabled, display msg provided.
  """
  if (on):
    print msg

def debugOut(msg, name=None, on=DEBUG):
  """
  When debug is enabled, output is printed to the display.
  """
  if name is not None:
    name = name+': '
  else:
    name = ''
  try:
    if on:
      print 'DEBUG: %s%s' % (name, msg)
    else:
      pass
  except:
    print 'MALFORMED DEBUG: %s%s' % (name, msg)

def errorOut(msg, ret=1):
  """
  When debug is enabled, output is printed to the display.

  returns nothing
  """
  if ret >= 1:
    prefix = 'ERROR'
  else:
    prefix = 'WARNING'
  out = '%s: %s' % (prefix, msg)
  if ret == 0:
    print out
  elif ret >= 1:
    out += '\n'
    sys.stderr.write(out)
    cleanUp(ret)

def cleanUp(ret):
  """
  Do a clean and proper sys.exit()

  executes sys.exit(int(ret))
  """
  sys.exit(ret)


"""
 Define Primary Functions
"""
def isHostsFileHashed(verbose=False):
  """
  Check to see if we can even get useful data out of the known_hosts files

  returns bool()
  """
  for config in SSH_CONFIGS:
    if os.path.exists(config) and not os.access(config, os.R_OK):
      raise KeployError, 'Unable to read config file %s' % (config)
    execute = "grep HashKnownHosts %s | awk '{print $2}'" % (config)
    debugOut(execute, '\tExecuting', verbose)
    is_hash = os.popen(execute).read().strip()
    debugOut(is_hash, '\tIs known_hosts hashed', verbose)
    if is_hash == "yes":
      return True
    return False

def getHostsFromFile(host_files, verbose=False):
  """
  Read hosts from provided file(s).

  returns list(hosts)
  """
  debugOut('getHostsFromFile(host_files=%s, verbose=%s)' % (host_files,
      verbose))
  if not isinstance(host_files, (list, tuple)):
    host_files = [host_files]
  hosts = []
  for host_file in host_files:
    if not os.path.exists(host_file):
      continue
    elif not os.access(host_file, os.R_OK):
      raise KeployError, 'Unable to parse hosts from file: %s' % (host_file)
    for line in open(host_file).readlines():
      host = line.strip().split()[0]
      if ',' in host:
        host = host.split(',')[0]
      if host not in hosts:
        debugOut(host, '\tFound Host')
        hosts.append(host)
  return list(hosts)

def getIdentity(id_file, verbose=False):
  """
  Read the identity file into a variable so that it is easier to push
  to the external host(s)

  returns str(identity)
  """
  debugOut('getIdentity(id_file=%s, verbose=%s)' % (id_file, verbose))
  if os.access(id_file, os.R_OK):
    identity = open(id_file).readlines()[0].strip()
    standardOut('\tProcessed identity:\n\t\t%s' % (id_file), verbose)
    debugOut(identity, '\tIdentity')
    return str(identity)
  raise KeployError('Could not find/access identity file: %s' % (id_file))

def findDefaultIdentityFile(id_files, verbose=False):
  """
  Return the first identity file name that is available from the provided list

  returns str(id_file)
  """
  debugOut('findDefaultIdentity(id_files=%s, verbose=%s)' % (id_files, verbose))
  for id_file in id_files:
    debugOut(id_file, '\tTrying to grab identity from')
    if os.access(id_file, os.R_OK):
      standardOut('\tFound identity file:\n\t\t%s' % (id_file), verbose)
      return str(id_file)
  raise KeployError('Could not find/access default identity files')

def toggleAgentForwarding(on, host, login_name, accept_key=False,
    timeout=None, verbose=False):
  """
  This function checks the state of ForwardAgent in remote system's
  ~/.ssh/config file, and toggles it, on/off as necessary

  returns bool(status)
  """
  ssh_options = []
  if isinstance(timeout, int):
    ssh_options.append(SSH_OPTION_TIMEOUT % (timeout))
  if accept_key:
    ssh_options.append(SSH_OPTION_ACCEPTKEY)
  ssh_call = SSH_START_CALL % (login_name, host, " ".join(ssh_options))
  command = ''
  forward_option = 'ForwardAgent'
  command +=  'grep -v \"%s\" %s > %s 2> /dev/null;' % (
    forward_option, SSH_CONFIGS[1], TMP_FILE)
  command += 'mv -f %s %s 2>/dev/null; chmod 600 %s 2>/dev/null;' % (
    TMP_FILE, SSH_CONFIGS[1], SSH_CONFIGS[1])
  if on:
    command += 'echo \'%s yes\' >> %s 2> /dev/null; grep \"%s\" %s' % (
      forward_option, SSH_CONFIGS[1], forward_option, SSH_CONFIGS[1])
  remote_command=ssh_call+command+SSH_END_CALL
  debugOut(remote_command, 'Executing')
  ret = os.popen(remote_command).readlines()
  if ret:
    if on:
      status = (True, 'enabled')
    else:
      status = (False, 'failed to disable')
    for line in ret:
      debugOut(line)
  else:
    if on:
      status = (False, 'failed to enable')
    else:
      status = (True, 'disabled')
  return status
  #standardOut('\t\tAgent Forwarding: %s' % (status), verbose)

def pushToRemoteHosts(hosts, identity, login_name, forward=False,
    remove_old=False, old_identity=None, accept_key=False,
    timeout=None, verbose=False):
  if not isinstance(hosts, (tuple, list)):
    hosts = list(hosts)
  for host in hosts:
    command = buildSSHPushCommand(host, identity, login_name, forward,
        remove_old, old_identity, accept_key, timeout, verbose)
    debugOut(command, 'Executing')
    ret = os.popen(command).readlines()
    if ret:
      if old_identity:
        status = 'changed'
      elif remove_old:
        status = 'failed to remove'
      else:
        status = 'deployed'
      for line in ret:
        debugOut(line)
    else:
      if remove_old:
        status = 'removed'
      else:
        status = 'failed to deploy'
    standardOut('\t\tPublic Identity Key: %s' % (status), verbose)

    if forward and not remove_old:
      toggleAgentForwarding(True, host, login_name, accept_key, timeout, verbose)

def buildSSHPushCommand(host, identity, login_name, forward=False,
      remove_old=False, old_identity=None, accept_key=False, timeout=None,
      verbose=False):
    ssh_options = []
    if isinstance(timeout, int):
      ssh_options.append(SSH_OPTION_TIMEOUT % (timeout))
    if accept_key:
      ssh_options.append(SSH_OPTION_ACCEPTKEY)
    ssh_call = SSH_START_CALL % (login_name, host, " ".join(ssh_options))
    standardOut('\tWorking on host: %s' % (host), verbose)
    if forward:
      toggleAgentForwarding(False, host, login_name, accept_key, timeout, verbose)
    command = ''
    if old_identity is None:
      old_identity = identity
    # Make ssh home dir and set permissions
    command += 'mkdir %s &> /dev/null; chmod 700 %s;' % (SSH_DIR, SSH_DIR)
    # Remove identity from file and set permissions
    command +=  'grep -v \"%s\" %s > %s 2> /dev/null;' % (
      old_identity, AUTH_KEYS_FILE, TMP_FILE)
    command += 'mv -f %s %s 2>/dev/null;' % (TMP_FILE, AUTH_KEYS_FILE)
    command += 'chmod 600 %s 2> /dev/null;' % (AUTH_KEYS_FILE)
    if not remove_old:
      # Then append the pub identity into the authorized_keys file, and grep it
      command += 'echo \'%s\' >> %s 2> /dev/null; chmod 600 %s; grep \"%s\" %s' % (
        identity, AUTH_KEYS_FILE, AUTH_KEYS_FILE, identity, AUTH_KEYS_FILE)
    return ssh_call+command+SSH_END_CALL


keploy/keploy.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# keploy - SSH Public Key Deplyment Utility
# Copyright © 2007 Greg Swift gregswift@gmail.com
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os
import signal
import sys

"""
 Define Variables
"""
DEBUG = True
VERSION = "0.5"

PW_WARN = '\nNOTICE: You may be prompted for you password,\n'
PW_WARN += 'NOTICE: this is directly from the ssh client, not keploy\n'

class KeployVariables:
  def __init__(self, user=None):
    self.ssh_bin = findSSHbin()
    self.ssh_config = '/etc/ssh/ssh_config'
    if (not os.access(self.ssh_config, os.R_OK)):
      self.ssh_config = None
    self.ssh_home_dir = os.path.expanduser('~/.ssh')
    self.tmp_file = os.path.join(self.ssh_home_dir, 'keyploy.tmp')
    self.id_files = (os.path.join(self.ssh_home_dir, 'id_rsa.pub'),
                     os.path.join(self.ssh_home_dir,'id_dsa.pub'),
                     os.path.join(self.ssh_home_dir,'identity.pub'))
    self.host_files = [os.path.join(self.ssh_home_dir,'known_hosts'),
                       os.path.join(self.ssh_home_dir,'known_hosts2')]
    self.auth_keys_file = os.path.join(self.ssh_home_dir,'authorized_keys')
    self.user = user

  def set(self, name, value):
    setattr(self, name, value)

"""
 Define Output Functions and cleanUp Function
"""
def standardOut(msg, on=True):
  """When quiet is not enabled, display msg provided """
  if (on):
    print msg

def debugOut(msg, name=None, on=DEBUG):
  """When debug is enabled, output is printed to the display."""
  if (name is not None):
    name = name+': '
  else:
    name = ''
  try:
    if (on):
      print 'DEBUG: %s%s' % (name, msg)
    else:
      pass
  except:
    print 'MALFORMED DEBUG: %s%s' % (name, msg)

def errorOut(msg, ret=1):
  """When debug is enabled, output is printed to the display.

 returns nothing"""
  type = ['WARNING', 'ERROR']
  out = '%s: %s' % (type[ret], msg)
  if (ret == 0):
    print out
  elif (ret == 1):
    out += '\n'
    sys.stderr.write(out)
    cleanUp(ret)

def cleanUp(ret):
  """Do a clean and proper sys.exit()

 executes sys.exit(int(ret))"""
  sys.exit(ret)


"""
 Define Primary Functions
"""
def getSSHbinary():
  """
  Try to locate the ssh binary using which.
  If 'None' we found nothing
  If 'False' it is not executeable (shouldn't happen, but why not test?)

  returns str(ssh_bin)
	"""
  ssh_bin = os.popen('which ssh').read().strip()
  if (ssh_bin == ''):
    ssh_bin = None
  elif (os.access(ssh_bin, os.X_OK)):
    ssh_bin = False
  return ssh_bin

def getHostsFromFile(get_from=None, known=False, verbose=True):
  """
 Read hosts from provided file.  If no file specified,
 try to read ~/.ssh/known_hosts

 returns list(hosts)
 """
  debugOut('getHostsFromFile(get_from=%s, known=%s, verbose=%s)' % (get_from,
      known, verbose))
  global host_files
  if (get_from is not None):
    if (os.access(get_from, os.R_OK)):
      standardOut('Using user-defined host list from file: %s' % (
        get_from), verbose)
      host_files = [get_from]
  else:    
    if (known):
      # First check to see if we can even get useful data out of the 
      # known_hosts file
      execute = "grep HashKnownHosts %s | awk '{print $2}'" % (ssh_config)
      debugOut(execute, '\tExecuting')
      is_hash = os.popen(execute).read().strip()
      debugOut(is_hash, '\tIs known_hosts hashed')
      if (is_hash == "yes"):
        msg = "The known_hosts files are hashed.  Please "
        msg += "specify host or alternate file to parse at cli"
        errorOut(msg)
      for f in host_files:
        try:
          os.path.exists(f)
        except:
          pass
        else:
          host_files.remove(f)
  hosts = []
  for x in host_files:
    execute = 'cat %s | cut -f1 -d" " | sed -e "s/,.*//g" | uniq' % (x)
    debugOut(execute, '\tExecuting')
    si, so, se = os.popen3(execute)
    for h in so.readlines():
      h = h.strip()
      debugOut(h, '\tFound Host')
      hosts.append(h)
  return list(hosts)

def getIdentity(id_file=None, verbose=True):
  """Read the identity file into a variable so that it is easier to push
to the external host(s)

 returns str(identity)"""
  debugOut('getIdentity(id_file=%s, verbose=%s)' % (id_file, verbose))
  global id_files
  if (id_file is not None):
    debugOut(id_file, '\tOver-riding id_files with')
    over_ride = True
    id_files = [os.path.abspath(os.path.expanduser(id_file))]
    standardOut('\tReading identity from:\n\t\t%s' % (id_file), verbose)
  for id_file in id_files:
    debugOut(id_file, '\tTrying to grab identity from')
    if (os.access(id_file, os.R_OK)):
      identity = os.popen('head -1 %s' % (id_file)).read().strip()
      msg = ''
      break
    else:
      if (over_ride):
        msg = 'Could not find/access specified file, %s' % (id_file)
      else:
        msg = 'Could not find/access default identity files'
  if (msg is ''):
    standardOut('\tFound identity:\n\t\t%s' % (id_file), verbose)
    debugOut(identity, '\tUsing identity')
  else:
    errorOut(msg)
  return str(identity)

def toggleAgentForwarding(on, ssh_call, end_ssh_call, verbose=True):
  """
 This function 
 """
  command = ''
  forward_option = 'ForwardAgent'
  ssh_user_config_file = os.path.join(ssh_home_dir, 'config')
  command +=  'grep -v \"%s\" %s > %s 2> /dev/null;' % (
    forward_option, ssh_user_config_file, tmp_file)
  command += 'mv -f %s %s 2>/dev/null; chmod 600 %s 2>/dev/null;' % (
    tmp_file, ssh_user_config_file, ssh_user_config_file)
  if (on):
    command += 'echo \'%s yes\' >> %s 2> /dev/null; grep \"%s\" %s' % (
      forward_option, ssh_user_config_file, forward_option, ssh_user_config_file)
  remote_command=ssh_call+command+end_ssh_call
  debugOut(remote_command, 'Executing')
  ret = os.popen(remote_command).readlines()
  if (ret):
    if (on):
      status = (True, 'enabled')
    else:
      status = (False, 'failed to disable')
    for line in ret:
      debugOut(line)
  else:
    if (on):
      status = (False, 'failed to enable')
    else:
      status = (True, 'disabled')
  return status
  #standardOut('\t\tAgent Forwarding: %s' % (status), verbose)

def pushToRemote(vars, options, hosts):
  for host in hosts:
    ssh_call = '%s -qq %s %s \'' % (vars.ssh_bin, options.userid, host)
    end_ssh_call = '\''
    standardOut('\tWorking on host: %s' % (host),
                options.verbose)
    if (options.forward):
      toggleAgentForwarding(False, ssh_call, end_ssh_call, options.verbose)
    command = ''
    try:
      old_identity
    except:
      use_identity = identity
    else:
      use_identity = old_identity
    # Make ssh home dir and set permissions
    command += 'mkdir %s &> /dev/null; chmod 700 %s;' % (
      ssh_home_dir, ssh_home_dir)
    # Remove identity from file and set permissions
    command +=  'grep -v \"%s\" %s > %s 2> /dev/null;' % (
      identity, auth_keys_file, tmp_file)
    command += 'mv -f %s %s 2>/dev/null;' % (tmp_file, auth_keys_file)
    command += 'chmod 600 %s 2> /dev/null;' % (auth_keys_file)
    if (not options.remove or options.old_id_file):
      use_identity = identity
      # Then append the pub identity into the authorized_keys file, and grep it
      command += 'echo \'%s\' >> %s 2> /dev/null; chmod 600 %s; grep \"%s\" %s' % (
        use_identity, auth_keys_file, auth_keys_file, use_identity, auth_keys_file)
    remote_command=ssh_call+command+end_ssh_call
    debugOut(remote_command, 'Executing')
    ret = os.popen(remote_command).readlines()
    if (ret):
      if (options.old_id_file):
        status = 'changed'
      elif (options.remove):
        status = 'failed to remove'
      else:
        status = 'deployed'
      for line in ret:
        debugOut(line)
    else:
      if (options.remove):
        status = 'removed'
      else:
        status = 'failed to deploy'
    standardOut('\t\tPublic Identity Key: %s' % (status), options.verbose)

    if (options.forward):
      if (not options.remove or options.old_id_file):
        toggleAgentForwarding(True, ssh_call, end_ssh_call, options.verbose)
